File: sort/shell_sort.cpp
#include <vector>
#include <iostream>

using namespace std;

void printNums(const vector<int>& nums, const char* prefix = "") {
    cout << prefix << "[ ";
    for (int num : nums) {
        cout << num << ", ";
    }
    cout << "]" << endl;
}

void shellSort(vector<int>& nums) {
    int delta = nums.size() / 2;
    while (delta != 0) {
        for (int i = delta; i < nums.size(); i++) {
            int j = i;
            int tmp = nums[i];
            while (j >= delta && tmp < nums[j - delta]) {
                nums[j] = nums[j - delta];
                j -= delta;
            }
            nums[j] = tmp;
            printNums(nums);
        }
        delta = delta / 2;
        printNums(nums, "after bound: ");
    }
}

int main(int argc, char const *argv[])
{
    vector<int> nums{7, 6, 5, 4, 3, 2, 1};
    shellSort(nums);    
    return 0;
}

File: sort/shell_sort.py

def shellSort(nums: list):
    n = len(nums)
    gap = n // 2
    while gap > 0:
        for i in range(gap, n):
            x = nums[i]
            j = i
            while j >= gap:
                if x < nums[j - gap]:
                    nums[j] = nums[j - gap]
                else:
                    break
                j -= gap
            nums[j] = x
        print(nums)
        gap = gap // 2
if __name__ == "__main__":
    nums = [10, 3, 1, 8, 4, 7, 6]
    shellSort(nums)
